Title: Methylation variations in Zea

# Collaborative Workflow

Jinliang: I suggest we following this [**workflow**](http://www.eqqon.com/index.php/Collaborative_Github_Workflow) for the project. Basically, you need to:
- **Fork** the repo
- **Clone** to your own server
- Do what ever you want in your forked repo, like **add**, **commit**, **push** without worry about the conflict issues
- Send **pull request** to the master repo **RILAB/GBSMethylation** when you want to share your progress
- If no conflict, we can just **merge** it. Otherwise, some body have to fix the conflicts
- Do not forget to **merge** from the updated RILAB master repo to your own fork
- Suggestion welcomed

# Brain storm
The md files are located in the folder `example_ideaKC/`
- [Kate's](example_ideaKC/KC_ideas.md)
- [Jinliang's](example_ideaKC/Jinliang.md)
- [Simon's]()


# Data:

3. Bisulfite sequencing of 20 teosinte plants
  	- Whole genome sequencing (~20X) and WGBS (~10-15X) on each
  	- GBS on ~100 progenies from each plant, all 100 phenotyped for multiple traits
  	- should be able to impute whole genome for progeny 
	- plants are from Palmar Chico. For more info on this population, see e.g. [our paper on fine-scale genetic stcuture](http://www.rilab.org/pdfs/Van-Heerwaarden-et-al-2010-951.pdf)
4. Bisulfite sequecing of ~20 addditional maize/teosinte  
	- TIL1 and TIL11 (unsure about access permission)
	- ~20 maize lines, including Oh43, Mo17, W22, and a diverse set of inbreds
5. Bob Schmitz mentioned possibility (informally) to sequence more
	- It would be interesting to see the [m6A methylome](http://www.nature.com/ncomms/2014/141128/ncomms6630/pdf/ncomms6630.pdf) in maize and teosinte.   


# Key points
1. Identification of methylation variation in Zea
2. Correlation with TE  
  ~~- Is there a TE map? [**DB of TE annotation is here**](http://maizetedb.org/~maize/)~~  
    - MC Stitzer re-annotation of TEs in maize genome
3. Whether selection sweeps associated with differential methylation patterns: see new [Tajima's D-like statistic](http://gbe.oxfordjournals.org/content/early/2014/12/22/gbe.evu271.short?rss=1&utm_source=dlvr.it&utm_medium=twitter)
4. Methylation and fitness?
5. Mapping the genetic control of methylation variation?
	* needs methylation data in offspring, which we don't have
	
See also recent example of WGBS in natural pops  [here](http://mbe.oxfordjournals.org/content/early/2014/12/21/molbev.msu344.abstract).
