KC ideas for 20 WGS and WGBiS Parents and current year 1 (50) GBS/Phenotype Offspring
===================================================================================

Most of my ideas are contingent on having WGBiS data available for the 50 year 1 offspring. These might need to be modified according to how I have interpreted the biology.

## Idea 1: comparison of methylation patterns between individuals with more homozygous vs. heteorzygous 

- Comparison 1 within the parental generation - prediction/thought: outbred individuals might have less methylated regions total (either in Mb or blocks) than homozygous individuals in the same generation. Why? Because if most epi-alleles are the obligate type (hypothesize that they are) and maize/teosinte has lots of crazy paralogs in it, than recombination ought to kick most of these out of the most outbred individuals. If most epi-alleles are the faciliated type in the population - then you may not see a difference at all between outbred vs. inbred individuals.

- Comparison 2 within the year 1 offspring (contingent on getting WGBiS data as I don't think methylation patterns in offspring would be easily imputed with just GBS): those progeny that have been selfed vs outcrossed - same hypothesis above - just a different way of getting at it?

- Note - I think there might be some statistical power issues here - unsure.

## Idea 2: differences in methylated regions in offspring according to who is mom dad (again contingent on Year 1 offspring having that WGBiS seqs) 

- There is evidence to suggest that inheritance patterns of genomic imprinting (methylated genes) may differ between female vs. male gametes. If we know the pollen donor and seed parent (and WGBiS sequences) of each of 50 offspring (and have the WGBiS data for this year 1 cohort), we might address this hypothesis? I'm more shaky on the biology here. Have to think a tad more. I know that stress can certainly alter these patterns. **What environment are Year 1 progeny grown in relative to the parents?**

## Idea 3: knobs in teosinte/maize and methylation
- If teosinte has lots of knob variants and there is a db out there characterizing these variants, is there an positive or negative association with knob sequences methylated regions of the genome? We could also look at this in maize (4 WGBiS)... I am assuming we can identify knobs based on sequence variation.
I would hypothesize that if knob variants are in fact "selfish" they ought not to be silenced - they should be non-methylated. 

## Idea 4: Genome size variation and methylation
- If there is genome size variation in a population of teosinte (not saying there's much here - but could easily get that with flow) - then perhaps we could look at whether this influences methylation. I'm not so excited by this - it's kind of tacked on. Guess I'm thinking large genomes have more non-coding DNA and more repetitive elements - if these are duplicated genes - perhaps more methylation, if not (non-coding and not of consequence) perhaps less. Think there'd be little power to do this type of comparison. Not sure of whether there's an association between "coding regions" and methylation - so I really have no idea what to expect here.

