---
##Methylation and CNVs/PAVs##

We have Bisulfite Seq for 20 teosinte from Palmar Chico Mexico, these samples are also being WGS to ~25x. I have also been calling PAV and CNVs in these twenty lines. As Jinliang points out in his suggestions we can map DMRs etc using established pipelines. 

My interest relates the whole genome duplication (WGD) and these data give us a nice opportunity to ask some novel biological question about epigenetics, PAVs/CNVs and ancient WGD.

**Concerning ancient WGD**

1. Do the different sub-genomes of *Zea* show evidence of varying methylation patterns (the sub-genomes have been well described and we can relatively easily answer this question provided we have nice estimates of methylation status across the genome).
1. How does this variation relate to rates of gene loss and the differences in gene expression that exist in the two sub-genomes (might be tough seeing as we don't have RNA-seq for the wild teosintes).
2. In terms of methlyation is there more varation in one of the sub-genomes relative to the other, at the population level?
3. Is the variation we see present in both wild (i.e. Palmar Chico pop) and the domesticated maize veriaties? 
4. Where are the DMRs? Intron exon, intergenic, repeat etc etc and does this vary between sub-genomes.

**Concerning PAVs/CNVs**

1.  Are methylation patterns over PAVs, CNVs differnt to the rest of the genome (I'm not quite sure of the detail here because my epigenetic knowledge is great).
2.  In the Palmar Chico pop do CNVs/PAVs leave a "methylation signature". For example are PAVs loci more likely to show more differential methylation when compared with a non-PAV genomic region. 
